sway/input/seatop_default.c

```diff
@@ -753,215 +752,322 @@ static void handle_pointer_axis(struct sway_seat *seat,
-	uint32_t button_end = 0;
-	if (button_base != 0) {
-		button_end = button_base + SWAY_SWIPE_DIR_COUNT;
-	}
-
-	/**
-	 * Handle mouse bindings - if there's any swipe bindings for this
-	 * context. Here we must swallow all swipes if there's even one
-	 * possible binding because we have no way of knowing yet what kind of
-	 * swipe this is going to be. It's in the nature of the thing because
-	 * the user may turn a left into an upwards swipe mid-way. For the same
-	 * reason it makes no sense to execute any binding here. So we defer it
-	 * to swipe release. We can break as soon as it's confirmed that there
-	 * is at least one binding.
-	 */
-	struct sway_binding *binding = NULL;
-	for (uint32_t button = button_base; button < button_end; button++) {
-		state_add_button(e, button);
-		// check for an on-press binding
-		binding = get_active_mouse_binding(e,
-				config->current_mode->mouse_bindings,
-				modifiers, false, on_titlebar, on_border,
-				on_contents, on_workspace, dev_id);
-		if (!binding) {
-			// see if there's an on-release binding instead
-			binding = get_active_mouse_binding(e,
-					config->current_mode->mouse_bindings,
-					modifiers, true, on_titlebar,
-					on_border, on_contents, on_workspace,
-					dev_id);
+		// If we already have a match ...
+		if(current) {
+			// ... check if input matching is equivalent
+			if(strcmp(current->input, binding->input) == 0) {
+
+				// ... - do not override an exact binding
+				if(!exact && current->flags & BINDING_EXACT) {
+					continue;
+				}
+
+				// ... - and ensure direction matching is better or equal
+				if (gesture_compare(&current->gesture, &binding->gesture) > 0) {
+					continue;
+				}
+			} else if(strcmp(binding->input, "*") == 0) {
+				// ... do not accept worse input match
+				continue;
+			}
 		}
 
-		state_erase_button(e, button);
-		if (binding) {
-			handled = true;
+		// Accept newer or better match
+		current = binding;
+
+		// If exact binding and input is found, quit search
+		if (strcmp(current->input, input) == 0 &&
+				gesture_compare(&current->gesture, gesture) == 0) {
 			break;
 		}
+	} // for all gesture bindings
+
+	return current;
+}
+
+// Wrapper around gesture_tracker_end to use tracker with sway bindings
+static struct sway_gesture_binding* gesture_tracker_end_and_match(
+		struct gesture_tracker *tracker, struct sway_input_device* device) {
+	// Determine name of input that received gesture
+	char *input = device
+		? input_device_get_identifier(device->wlr_device)
+		: strdup("*");
+
+	// Match tracking result to binding
+	struct gesture *gesture = gesture_tracker_end(tracker);
+	struct sway_gesture_binding *binding = gesture_binding_match(
+			config->current_mode->gesture_bindings, gesture, input);
+	free(gesture);
+	free(input);
+
+	return binding;
+}
+
+/// Small wrapper around seat_execute_command to work on gesture bindings
+static void gesture_binding_execute(struct sway_seat *seat,
```

Comment:
Just reading my old comments here, and I think such a flag should not be included. If there is a desire for it we can add it later.

Reply:
Ok, I will have another look, just so I fully understand this. And we receive a surface when a gesture begins, so it does make sense to use that information. I won't spent to much time on this.

Reply:
So it seems the main difference is that mouse bindings provide `execute_command` with an additional container pointer in `seat_execute_command`. It seems however the context of this container is only relevant, if you do not provide any criteria, so I am not sure your example applies, or ever stopped working, because criteria are handled in `execute command` either way. So I added that same behavior for gesture events. I did not use the surface passed to at the beginning of a gesture, and I think that is more an issue for another PR.